File: novo-mockup/_shared/shell.js
/* Ouroboros — shell (sidebar + topbar) renderizado por JS.
   Cada mockup chama montarShell({clusterAtivo, abaAtiva, breadcrumb}). */

const CLUSTERS_OUROBOROS = [
  {
    id: 'inbox',
    nome: 'Inbox',
    glyph: 'inbox',
    badge: '3',
    abas: [
      { id: 'inbox', nome: 'Inbox', file: '16-inbox.html' },
    ],
  },
  {
    id: 'home',
    nome: 'Home',
    glyph: 'home',
    abas: [
      { id: 'visao-geral', nome: 'Visão Geral', file: '01-visao-geral.html' },
      { id: 'home-financas', nome: 'Finanças', file: '02-extrato.html' },
      { id: 'home-documentos', nome: 'Documentos', file: '06-busca-global.html' },
      { id: 'home-analise', nome: 'Análise', file: '12-analise.html' },
      { id: 'home-metas', nome: 'Metas', file: '13-metas.html' },
    ],
  },
  {
    id: 'financas',
    nome: 'Finanças',
    glyph: 'financas',
    abas: [
      { id: 'extrato', nome: 'Extrato', file: '02-extrato.html' },
      { id: 'contas', nome: 'Contas', file: '03-contas.html' },
      { id: 'pagamentos', nome: 'Pagamentos', file: '04-pagamentos.html' },
      { id: 'projecoes', nome: 'Projeções', file: '05-projecoes.html' },
    ],
  },
  {
    id: 'documentos',
    nome: 'Documentos',
    glyph: 'docs',
    abas: [
      { id: 'busca-global', nome: 'Busca Global', file: '06-busca-global.html' },
      { id: 'catalogacao', nome: 'Catalogação', file: '07-catalogacao.html' },
      { id: 'completude', nome: 'Completude', file: '08-completude.html' },
      { id: 'revisor', nome: 'Revisor', file: '09-revisor.html' },
      { id: 'validacao-arquivos', nome: 'Validação por Arquivo', file: '10-validacao-arquivos.html' },
      { id: 'grafo-obsidian', nome: 'Grafo + Obsidian', file: '06-busca-global.html' },
    ],
  },
  {
    id: 'analise',
    nome: 'Análise',
    glyph: 'analise',
    abas: [
      { id: 'categorias', nome: 'Categorias', file: '11-categorias.html' },
      { id: 'analise', nome: 'Análise', file: '12-analise.html' },
      { id: 'irpf', nome: 'IRPF', file: '15-irpf.html' },
    ],
  },
  {
    id: 'metas',
    nome: 'Metas',
    glyph: 'metas',
    abas: [
      { id: 'metas', nome: 'Metas', file: '13-metas.html' },
    ],
  },
  {
    id: 'bem-estar',
    nome: 'Bem-estar',
    glyph: 'heart',
    abas: [
      { id: 'be-hoje',     nome: 'Hoje',             file: '17-bem-estar-hoje.html' },
      { id: 'be-humor',    nome: 'Humor',            file: '18-humor-heatmap.html' },
      { id: 'be-diario',   nome: 'Diário emocional', file: '19-diario-emocional.html' },
      { id: 'be-rotina',   nome: 'Rotina',           file: '20-rotina.html' },
      { id: 'be-recap',    nome: 'Recap',            file: '21-recap.html' },
      { id: 'be-eventos',  nome: 'Eventos',          file: '22-eventos.html' },
      { id: 'be-memorias', nome: 'Memórias',         file: '23-memorias.html' },
      { id: 'be-medidas',  nome: 'Medidas',          file: '24-medidas.html' },
      { id: 'be-ciclo',    nome: 'Ciclo',            file: '25-ciclo.html' },
      { id: 'be-cruzamentos', nome: 'Cruzamentos',   file: '26-cruzamentos.html' },
      { id: 'be-privacidade', nome: 'Privacidade A↔B', file: '27-privacidade.html' },
      { id: 'be-rotina-toml', nome: 'Editor rotina.toml', file: '28-rotina-toml.html' },
    ],
  },
  {
    id: 'sistema',
    nome: 'Sistema',
    glyph: 'terminal',
    abas: [
      { id: 'skills-d7', nome: 'Skills D7', file: '14-skills-d7.html' },
      { id: 'styleguide', nome: 'Styleguide', file: '../styleguide.html' },
      { id: 'indice', nome: 'Índice', file: '../index.html' },
    ],
  },
];

function _resolveHref(file) {
  // Detecta se estamos em /mockups/ ou na raiz, e ajusta o prefixo.
  const naRaiz = !window.location.pathname.includes('/mockups/');
  if (file.startsWith('../')) {
    // file aponta para raiz a partir de mockups/
    return naRaiz ? './' + file.slice(3) : file;
  }
  // file é um mockup (ex: '14-skills-d7.html')
  return naRaiz ? './mockups/' + file : './' + file;
}

function _sidebarHTML(clusterAtivo, abaAtiva) {
  const clusters = CLUSTERS_OUROBOROS.map((c) => {
    const itens = c.abas.map((a) => {
      const ativa = (c.id === clusterAtivo && a.id === abaAtiva);
      const href = a.file ? _resolveHref(a.file) : '#';
      const cls = ativa ? 'sidebar-item active' : 'sidebar-item';
      const semMockup = !a.file ? '<span class="count" title="ainda não modelada">·</span>' : '';
      return `<a class="${cls}" href="${href}" data-aba="${a.id}">${a.nome}${semMockup}</a>`;
    }).join('');
    const badge = c.badge ? `<span class="badge">${c.badge}</span>` : '';
    return `
      <div class="sidebar-cluster">
        <div class="sidebar-cluster-header">
          <span style="display:inline-flex;align-items:center;gap:8px;">
            ${glyph(c.glyph, 14)} ${c.nome}
          </span>
          ${badge}
        </div>
        ${itens}
      </div>`;
  }).join('');

  return `
    <aside class="sidebar" aria-label="Navegação">
      <a class="sidebar-brand" href="${_resolveHref('../index.html')}" style="text-decoration:none;color:inherit;">
        ${glyph('ouroboros', 18)}
        <span>Ouroboros</span>
      </a>
      <div class="sidebar-search">
        ${glyph('search', 14, 'sidebar-search-icon')}
        <input type="text" placeholder="Buscar fornecedor, sha8, valor..." aria-label="Buscar">
        <kbd>/</kbd>
      </div>
      ${clusters}
      <div style="margin-top:auto;padding:12px 16px;border-top:1px solid var(--border-subtle);font-size:11px;color:var(--text-muted);font-family:var(--ff-mono);">
        <div>D7 · cobertura observável</div>
      </div>
    </aside>`;
}

function _topbarHTML(breadcrumb, acoes) {
  const segs = breadcrumb.map((s, i) => {
    const last = i === breadcrumb.length - 1;
    const cls = last ? 'seg current' : 'seg';
    const sep = last ? '' : '<span class="sep">/</span>';
    // primeiro segmento ("Ouroboros") leva ao índice
    if (i === 0 && !last) {
      return `<a class="${cls}" href="${_resolveHref('../index.html')}" style="text-decoration:none;color:inherit;">${s}</a>${sep}`;
    }
    return `<span class="${cls}">${s}</span>${sep}`;
  }).join('');
  const acoesHTML = (acoes || []).map((a) => {
    const cls = a.primary ? 'btn btn-primary btn-sm' : 'btn btn-sm';
    const inner = `${a.icon ? glyph(a.icon, 14) : ''} ${a.label}`;
    if (a.href) {
      const href = a.href.endsWith('.html') ? _resolveHref(a.href) : a.href;
      return `<a class="${cls}" href="${href}" style="text-decoration:none;display:inline-flex;align-items:center;gap:6px;">${inner}</a>`;
    }
    if (a.onClick) {
      return `<button class="${cls}" onclick="${a.onClick}">${inner}</button>`;
    }
    return `<button class="${cls}">${inner}</button>`;
  }).join('');
  return `
    <header class="topbar">
      <nav class="breadcrumb" aria-label="Localização">${segs}</nav>
      <div class="topbar-actions">${acoesHTML}</div>
    </header>`;
}

/**
 * Monta sidebar + topbar dentro de um <div id="shell-root"> e devolve <main>.
 * @param {Object} opts
 * @param {string} opts.clusterAtivo - id do cluster
 * @param {string} opts.abaAtiva - id da aba
 * @param {string[]} opts.breadcrumb - segmentos
 * @param {Array} opts.acoes - [{label, icon, primary}]
 */
function montarShell(opts) {
  const root = document.getElementById('shell-root');
  if (!root) return;
  root.classList.add('shell');
  root.innerHTML =
    _sidebarHTML(opts.clusterAtivo, opts.abaAtiva) +
    _topbarHTML(opts.breadcrumb, opts.acoes) +
    '<main class="main" id="main-root"></main>';
  // Atalhos de teclado: g h, g i, g v, g r, g f, /, ?
  _instalarAtalhos();
}

function _instalarAtalhos() {
  let buffer = '';
  let timer = null;
  const mapa = {
    'gh': _resolveHref('01-visao-geral.html'),
    'gi': _resolveHref('16-inbox.html'),
    'gv': _resolveHref('10-validacao-arquivos.html'),
    'gr': _resolveHref('09-revisor.html'),
    'gf': _resolveHref('15-irpf.html'),
    'gc': _resolveHref('07-catalogacao.html'),
  };
  document.addEventListener('keydown', (e) => {
    if (e.target.matches('input, textarea')) return;
    if (e.key === '/') {
      e.preventDefault();
      const inp = document.querySelector('.sidebar-search input');
      if (inp) inp.focus();
      return;
    }
    if (e.key === '?') {
      _toggleAjuda();
      return;
    }
    if (e.key === 'Escape') {
      const overlay = document.getElementById('ajuda-overlay');
      if (overlay) overlay.remove();
      return;
    }
    buffer += e.key.toLowerCase();
    clearTimeout(timer);
    timer = setTimeout(() => { buffer = ''; }, 800);
    if (mapa[buffer]) {
      const dest = mapa[buffer];
      buffer = '';
      window.location.href = dest;
    }
  });
}

/* Toast — feedback simples para ações sem backend.
   Uso: toast('Mensagem'); ou toast('Erro', 'error'); */
function toast(msg, tipo = 'ok') {
  let host = document.getElementById('toast-host');
  if (!host) {
    host = document.createElement('div');
    host.id = 'toast-host';
    host.style.cssText = 'position:fixed;bottom:24px;right:24px;z-index:80;display:flex;flex-direction:column;gap:8px;pointer-events:none;';
    document.body.appendChild(host);
  }
  const cor = tipo === 'error' ? 'var(--accent-red)' : tipo === 'warn' ? 'var(--accent-yellow)' : 'var(--accent-purple)';
  const t = document.createElement('div');
  t.style.cssText = `pointer-events:auto;background:var(--bg-elevated);border:1px solid var(--border-subtle);border-left:3px solid ${cor};border-radius:var(--r-sm);padding:10px 14px;font-family:var(--ff-mono);font-size:12px;color:var(--text-primary);box-shadow:0 8px 24px rgba(0,0,0,0.4);min-width:240px;opacity:0;transform:translateX(20px);transition:opacity .2s,transform .2s;`;
  t.textContent = msg;
  host.appendChild(t);
  requestAnimationFrame(() => { t.style.opacity = '1'; t.style.transform = 'translateX(0)'; });
  setTimeout(() => { t.style.opacity = '0'; t.style.transform = 'translateX(20px)'; setTimeout(() => t.remove(), 250); }, 2400);
}

function _toggleAjuda() {
  const existente = document.getElementById('ajuda-overlay');
  if (existente) { existente.remove(); return; }
  const ov = document.createElement('div');
  ov.id = 'ajuda-overlay';
  ov.style.cssText = 'position:fixed;inset:0;background:rgba(0,0,0,0.7);z-index:60;display:grid;place-items:center;';
  ov.innerHTML = `
    <div class="card" style="min-width:480px;max-width:560px;">
      <div class="card-head">
        <h3 class="card-title">Atalhos de teclado</h3>
        <button class="btn btn-icon btn-ghost" onclick="document.getElementById('ajuda-overlay').remove()">${glyph('close',14)}</button>
      </div>
      <table style="width:100%;font-size:13px;">
        <tr><td style="padding:6px 0;"><kbd class="kbd">g h</kbd></td><td>Visão Geral</td></tr>
        <tr><td style="padding:6px 0;"><kbd class="kbd">g i</kbd></td><td>Inbox</td></tr>
        <tr><td style="padding:6px 0;"><kbd class="kbd">g v</kbd></td><td>Validação por Arquivo</td></tr>
        <tr><td style="padding:6px 0;"><kbd class="kbd">g r</kbd></td><td>Revisor</td></tr>
        <tr><td style="padding:6px 0;"><kbd class="kbd">g f</kbd></td><td>IRPF</td></tr>
        <tr><td style="padding:6px 0;"><kbd class="kbd">g c</kbd></td><td>Catalogação</td></tr>
        <tr><td style="padding:6px 0;"><kbd class="kbd">/</kbd></td><td>Focar busca</td></tr>
        <tr><td style="padding:6px 0;"><kbd class="kbd">?</kbd></td><td>Esta ajuda</td></tr>
        <tr><td style="padding:6px 0;"><kbd class="kbd">Esc</kbd></td><td>Fechar overlay</td></tr>
      </table>
    </div>`;
  document.body.appendChild(ov);
  ov.addEventListener('click', (e) => { if (e.target === ov) ov.remove(); });
}

File: src/dashboard/componentes/shell.py
"""Shell global do dashboard: sidebar de 8 clusters + topbar — Sprint UX-RD-03.

Versão Python+Streamlit do contrato de ``novo-mockup/_shared/shell.js``
(funções ``_sidebarHTML`` e ``_topbarHTML``). Cada função emite um bloco
HTML pronto para ``st.markdown(html, unsafe_allow_html=True)`` e consome
exclusivamente classes definidas em ``src/dashboard/tema_css.py`` (UX-RD-02).

Topologia da sidebar (ordem canônica espelha ``CLUSTERS_VALIDOS`` em
``componentes/drilldown.py``):

  1. Inbox      -- fila de novos arquivos (UX-RD-15 implementa)
  2. Home       -- visão geral + 4 mini-views cross-area
  3. Finanças   -- Extrato, Contas, Pagamentos, Projeções
  4. Documentos -- Busca Global, Catalogação, Completude, Revisor,
                   Validação por Arquivo, Grafo + Obsidian
  5. Análise    -- Categorias, Análise, IRPF
  6. Metas      -- Metas
  7. Bem-estar  -- 12 telas pessoais não-financeiras (UX-RD-16+ implementa)
  8. Sistema    -- Skills D7, Styleguide, Índice (UX-RD-05 implementa)

Cada item usa link ``href="?cluster=<X>&tab=<Y>"`` para preservar o
deep-link da Sprint 100. A tab ``a class="sidebar-item active"`` recebe
o destaque visual via CSS já existente em ``tema_css.py`` (gradient +
``border-left 2px var(--accent-purple)``).

Segurança XSS: nomes de cluster, abas e segmentos do breadcrumb são
escapados via ``html.escape`` antes da concatenação. ``href`` usa apenas
componentes hardcoded em ``CLUSTERS_REDESIGN`` ou em ``ABAS_POR_CLUSTER``
de ``app.py`` (não há valores arbitrários em runtime).
"""

from __future__ import annotations

import html
import urllib.parse
from collections.abc import Iterable

# Estrutura canônica da sidebar redesenhada. Espelha 1:1 o array
# ``CLUSTERS_OUROBOROS`` de ``novo-mockup/_shared/shell.js``. Os nomes de aba
# coincidem com os labels usados em ``ABAS_POR_CLUSTER`` (em ``app.py``)
# para clusters que já têm páginas implementadas (Home, Finanças,
# Documentos, Análise, Metas). Os 3 clusters novos (Inbox, Bem-estar,
# Sistema) ainda não têm páginas em ``paginas/``; o dispatcher em
# ``main()`` mostra fallback graceful.
CLUSTERS_REDESIGN: tuple[dict, ...] = (
    {
        "nome": "Inbox",
        "abas": (
            {"nome": "Inbox", "implementada": False, "sprint_alvo": "UX-RD-15"},
        ),
    },
    {
        "nome": "Home",
        "abas": (
            {"nome": "Visão Geral", "implementada": True},
            {"nome": "Finanças", "implementada": True},
            {"nome": "Documentos", "implementada": True},
            {"nome": "Análise", "implementada": True},
            {"nome": "Metas", "implementada": True},
        ),
    },
    {
        "nome": "Finanças",
        "abas": (
            {"nome": "Extrato", "implementada": True},
            {"nome": "Contas", "implementada": True},
            {"nome": "Pagamentos", "implementada": True},
            {"nome": "Projeções", "implementada": True},
        ),
    },
    {
        "nome": "Documentos",
        "abas": (
            {"nome": "Busca Global", "implementada": True},
            {"nome": "Catalogação", "implementada": True},
            {"nome": "Completude", "implementada": True},
            {"nome": "Revisor", "implementada": True},
            {"nome": "Validação por Arquivo", "implementada": True},
            {"nome": "Grafo + Obsidian", "implementada": True},
        ),
    },
    {
        "nome": "Análise",
        "abas": (
            {"nome": "Categorias", "implementada": True},
            {"nome": "Análise", "implementada": True},
            {"nome": "IRPF", "implementada": True},
        ),
    },
    {
        "nome": "Metas",
        "abas": ({"nome": "Metas", "implementada": True},),
    },
    {
        "nome": "Bem-estar",
        "abas": (
            # Sprint UX-RD-17: Hoje e Humor estão implementadas; demais
            # 10 abas seguem com fallback graceful e sprint_alvo
            # apontando para a sprint que vai habilitar cada uma.
            {"nome": "Hoje", "implementada": True},
            {"nome": "Humor", "implementada": True},
            {"nome": "Diário", "implementada": False, "sprint_alvo": "UX-RD-18"},
            {"nome": "Eventos", "implementada": False, "sprint_alvo": "UX-RD-19"},
            {"nome": "Medidas", "implementada": False, "sprint_alvo": "UX-RD-20"},
            {"nome": "Treinos", "implementada": False, "sprint_alvo": "UX-RD-21"},
            {"nome": "Marcos", "implementada": False, "sprint_alvo": "UX-RD-22"},
            {"nome": "Alarmes", "implementada": False, "sprint_alvo": "UX-RD-23"},
            {"nome": "Contadores", "implementada": False, "sprint_alvo": "UX-RD-24"},
            {"nome": "Ciclo", "implementada": False, "sprint_alvo": "UX-RD-25"},
            {"nome": "Tarefas", "implementada": False, "sprint_alvo": "UX-RD-26"},
            {"nome": "Recap", "implementada": False, "sprint_alvo": "UX-RD-27"},
        ),
    },
    {
        "nome": "Sistema",
        "abas": (
            {"nome": "Skills D7", "implementada": False, "sprint_alvo": "UX-RD-05"},
        ),
    },
)


def _href_para(cluster: str, aba: str | None = None) -> str:
    """Monta query-string ``?cluster=...&tab=...`` com URL-encoding.

    Sempre devolve string que começa com ``?``. ``cluster`` é obrigatório;
    ``aba`` é omitida quando ``None``.
    """
    params: list[tuple[str, str]] = [("cluster", cluster)]
    if aba is not None:
        params.append(("tab", aba))
    return "?" + urllib.parse.urlencode(params)


def _renderizar_brand_html() -> str:
    """Bloco do brand (topo da sidebar): glyph SVG ouroboros + nome em mono uppercase.

    FIX-07: porta o glyph 'ouroboros' do mockup canônico (componentes/glyphs.py)
    em vez de placeholder letra "O". Antes: ``<span>O</span>`` (visualmente lia
    "OOuroboros"); agora: ``<svg viewBox="0 0 24 24">...</svg>`` herdando cor.
    """
    from src.dashboard.componentes.glyphs import glyph  # noqa: PLC0415

    return (
        '<a class="sidebar-brand" href="?cluster=Home" '
        'style="text-decoration:none;color:inherit;">'
        f'{glyph("ouroboros", tamanho_px=20, classe="sidebar-brand-glyph")}'
        "<span>Ouroboros</span>"
        "</a>"
    )


def _renderizar_busca_html() -> str:
    """Bloco do campo de busca placeholder + tecla `/`.

    Sprint UX-RD-03 usa o input HTML estático espelhando o mockup. O
    componente Streamlit interativo (``renderizar_input_busca`` da Sprint
    UX-113/114) continua sendo renderizado via Python por ``app.py`` --
    este input HTML é apenas decorativo, mostra o placeholder e o ``kbd``,
    e provê o alvo de foco para a tecla ``/`` mesmo antes do componente
    Streamlit hidratar.
    """
    return (
        '<div class="sidebar-search">'
        '<span class="sidebar-search-icon" aria-hidden="true">?</span>'
        '<input type="text" placeholder="Buscar fornecedor, sha8, valor..." '
        'aria-label="Buscar (atalho: tecla /)" '
        'data-ouroboros-busca="placeholder" />'
        "<kbd>/</kbd>"
        "</div>"
    )


def _renderizar_cluster_html(
    cluster: dict, cluster_ativo: str, aba_ativa: str
) -> str:
    """Bloco de um cluster: header + lista de itens (abas)."""
    nome_cluster = cluster["nome"]
    nome_seguro = html.escape(nome_cluster)

    itens_html: list[str] = []
    for aba in cluster["abas"]:
        nome_aba = aba["nome"]
        ativa = nome_cluster == cluster_ativo and nome_aba == aba_ativa
        # Quando o cluster é o ativo e nenhuma aba específica foi marcada,
        # destacamos a primeira aba implementada (default visual).
        classe = "sidebar-item active" if ativa else "sidebar-item"
        # FIX-12: aria-current="page" no item ativo para tecnologias
        # assistivas (WCAG 1.3.1 Info and Relationships).
        aria_current = ' aria-current="page"' if ativa else ""
        href = _href_para(nome_cluster, nome_aba)
        rotulo = html.escape(nome_aba)
        if not aba.get("implementada", True):
            sprint_alvo = aba.get("sprint_alvo", "")
            badge = (
                f'<span class="count" title="A implementar em {html.escape(sprint_alvo)}">'
                "...</span>"
            )
        else:
            badge = ""
        itens_html.append(
            f'<a class="{classe}"{aria_current} href="{href}" data-cluster="{nome_seguro}" '
            f'data-aba="{rotulo}">{rotulo}{badge}</a>'
        )

    # UX-U-01: badge canônica no header do cluster (mockup 00-shell-navegacao
    # mostra <span class="badge">3</span> no Inbox). Por hora apenas Inbox
    # recebe badge; valor placeholder "..." enquanto contagem real do
    # data/raw/_inbox/ não está implementada.
    badge_cluster_html = ""
    if nome_cluster == "Inbox":
        badge_cluster_html = '<span class="badge" title="Arquivos pendentes na fila">...</span>'

    return (
        '<div class="sidebar-cluster">'
        '<div class="sidebar-cluster-header">'
        f'<span style="display:inline-flex;align-items:center;gap:8px;">'
        f"{nome_seguro}</span>"
        f"{badge_cluster_html}"
        "</div>"
        + "".join(itens_html)
        + "</div>"
    )


def renderizar_sidebar(cluster_ativo: str, aba_ativa: str = "") -> str:
    """Devolve o HTML completo da sidebar redesenhada (8 clusters).

    Args:
        cluster_ativo: nome do cluster atualmente ativo (ex: "Home"). Se
            o nome não estiver em ``CLUSTERS_REDESIGN``, nenhum item é
            destacado (graceful).
        aba_ativa: nome da aba dentro do cluster ativo. Se vazia, nenhuma
            aba específica é destacada.

    Returns:
        Bloco HTML pronto para ``st.markdown(html, unsafe_allow_html=True)``
        DENTRO de ``with st.sidebar:``.
    """
    blocos: list[str] = []
    # FIX-12: skip-link no início da sidebar (WCAG 2.4.1 Bypass Blocks).
    # Invisível por padrão; aparece quando recebe foco via Tab.
    blocos.append(
        '<a class="skip-link sr-only-focusable" href="#main-root">'
        "Pular para conteúdo principal"
        "</a>"
    )
    blocos.append(_renderizar_brand_html())
    blocos.append(_renderizar_busca_html())
    for cluster in CLUSTERS_REDESIGN:
        blocos.append(_renderizar_cluster_html(cluster, cluster_ativo, aba_ativa))
    blocos.append(
        '<div class="sidebar-footer" '
        'style="margin-top:auto;padding:12px 16px;'
        "border-top:1px solid var(--border-subtle);"
        "font-size:11px;color:var(--text-muted);"
        'font-family:var(--ff-mono);">'
        "<div>D7 - cobertura observável</div>"
        "</div>"
    )
    # ``aside.sidebar`` envolve o bloco completo. ``aria-label`` para AT.
    return (
        '<aside class="sidebar ouroboros-sidebar-redesign" aria-label="Navegação">'
        + "".join(blocos)
        + "</aside>"
    )


def renderizar_topbar(
    breadcrumb: Iterable[str], acoes: Iterable[dict] | None = None
) -> str:
    """Devolve o HTML completo da topbar com breadcrumb + slot de ações.

    Args:
        breadcrumb: segmentos do caminho atual (ex: ``["Ouroboros", "Home",
            "Visão Geral"]``). O último segmento recebe a classe ``current``
            (cor primary). Os demais ficam em ``text-secondary`` separados
            por ``/`` em ``border-strong``.
        acoes: iterável opcional de dicts ``{"label": str, "href": str|None,
            "primary": bool}``. Esta sprint emite apenas o slot vazio; sprints
            de página (UX-RD-04+) vão preencher.

    Returns:
        Bloco HTML pronto para ``st.markdown(html, unsafe_allow_html=True)``
        no início de ``main()``, antes do conteúdo principal.
    """
    # FIX-05: segmentos não-current viram <a href="?cluster=...">; current
    # permanece <span>. Mockup canônico (components.css:114-121) define
    # .breadcrumb .seg como elemento clicável; current é a página atual
    # (não-clicável). Importação atrasada de CLUSTERS_VALIDOS para evitar
    # ciclo. "Ouroboros" (primeiro segmento) aponta para ?cluster=Home.
    from src.dashboard.componentes.drilldown import CLUSTERS_VALIDOS  # noqa: PLC0415

    segmentos = list(breadcrumb)
    n = len(segmentos)
    partes_seg: list[str] = []
    for i, seg in enumerate(segmentos):
        ultimo = i == n - 1
        rotulo = html.escape(seg)
        sep = "" if ultimo else '<span class="sep">/</span>'
        if ultimo:
            partes_seg.append(f'<span class="seg current">{rotulo}</span>{sep}')
        else:
            # primeiro segmento "Ouroboros" -> Home; demais -> cluster homônimo se válido
            if i == 0:
                href_pai = "?cluster=Home"
            elif seg in CLUSTERS_VALIDOS:
                href_pai = f"?cluster={html.escape(seg, quote=True)}"
            else:
                href_pai = "?cluster=Home"
            partes_seg.append(
                f'<a class="seg" href="{href_pai}" '
                'style="text-decoration:none;color:inherit;">'
                f'{rotulo}</a>{sep}'
            )

    partes_acoes: list[str] = []
    for acao in acoes or []:
        label = html.escape(str(acao.get("label", "")))
        href = acao.get("href")
        classe = "btn btn-primary btn-sm" if acao.get("primary") else "btn btn-sm"
        if href:
            href_esc = html.escape(str(href), quote=True)
            partes_acoes.append(
                f'<a class="{classe}" href="{href_esc}" '
                'style="text-decoration:none;display:inline-flex;'
                'align-items:center;gap:6px;">'
                f"{label}</a>"
            )
        else:
            partes_acoes.append(f'<button class="{classe}">{label}</button>')

    # UX-U-02: slot dinâmico vindo de st.session_state['topbar_acoes_html'],
    # alimentado pelo helper componentes/topbar_actions.renderizar_grupo_acoes()
    # que cada página pode chamar no início de seu renderizar(). main() em
    # app.py reseta o slot antes de cada run para evitar leak entre páginas.
    acoes_session_html = ""
    try:
        import streamlit as _st
        acoes_session_html = _st.session_state.get("topbar_acoes_html", "")
    except Exception:
        acoes_session_html = ""

    slot_html = "".join(partes_acoes) + acoes_session_html

    return (
        '<header class="topbar ouroboros-topbar-redesign">'
        f'<nav class="breadcrumb" aria-label="Localização">{"".join(partes_seg)}</nav>'
        f'<div class="topbar-actions">{slot_html}</div>'
        "</header>"
    )


def instalar_atalhos_globais() -> None:
    """Injeta o JS de atalhos via ``st.components.v1.html(..., height=0)``.

    Idempotente: o script tem guard ``window.__ouroborosAtalhosInstalados``
    que impede múltiplos listeners. Pode ser chamada uma vez por execução
    de ``main()`` sem efeito colateral acumulativo.
    """
    try:
        from streamlit.components import v1 as components
    except ImportError:  # pragma: no cover -- streamlit indisponível em testes puros
        return

    from src.dashboard.componentes.atalhos_teclado import gerar_html_atalhos

    components.html(gerar_html_atalhos(), height=0)


# "Tudo flui, nada permanece." -- Heráclito
